### PR TITLE
[ENG-2014]  Browse node children when reftype HasChild is encountered

### PR DESCRIPTION
**Pull request by kanapuli**
## Description

RefType HasChild is supported while browsing opcua nodes. The browse code is refactored to mock its functionality and tests are added.

`browseHierarchicalReferences` setting is added to the opcUA plugin to browse all references of a node without any duplicates.

https://github.com/user-attachments/assets/ace8426b-46f3-4182-aeaa-017cadd94e5c

## Release Notes

- **New Features**
	- Introduced enhanced browsing functionality for OPC UA nodes, including a new configuration option to browse hierarchical references.
	- Added a new `NodeBrowser` interface for improved management of OPC UA nodes.
	- Enhanced logging capabilities with a new `Logger` interface for better monitoring.

- **Bug Fixes**
	- Improved error handling and logging during node browsing operations.
	- Enhanced error handling in the reading process to better manage session validity and connection issues.

- **Tests**
	- Expanded unit tests for the browsing functionality, including various scenarios for better reliability and clarity.

### CODE REVIEW

File: opcua_plugin/browse_test.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+		parentNodeId = ""
+		nodeChan = make(chan NodeDef, 100)
+		errChan = make(chan error, 100)
+		pathIDMapChan = make(chan map[string]string, 100)
+		wg = &sync.WaitGroup{}
+	})
+
+	Context("When browsing nodes with a node class value nil", func() {
+		It("should return an error for nil node class in the error channel", func() {
+			Skip("temp")
```
Comment (JeremyTheocharis) [inline opcua_plugin/browse_test.go:145]: temp?

File: opcua_plugin/browse_test.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+				errs = append(errs, err)
+			}
+
+			Expect(errs).Should(HaveLen(1))
+			Expect(errs[0].Error()).To(Equal("opcua: node class is nil"))
+		})
+
+		It("should browse the root node with no children with right attributes", func() {
+
+			rootNode := createMockVariableNode(1234, "TestNode")
```
Comment (JeremyTheocharis) [inline opcua_plugin/browse_test.go:179]: does this now result in having two times the same "browseName" attirburte? once through the createMockVariableNode and then once through appending an attribute with rootNode.attriburtes = append(..)?

File: opcua_plugin/browse_test.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
+package opcua_plugin
```
Comment (JeremyTheocharis) [inline opcua_plugin/browse_test.go:1]: I love this test. this will make things so much easier in the future to add new edge cases

**Review by JeremyTheocharis — CHANGES_REQUESTED** [review]

File: opcua_plugin/browse_test.go (hunk 1 above)
Comment (kanapuli) [reply opcua_plugin/browse_test.go:145]: Oh, I forgot remove it. Sorry.

File: opcua_plugin/browse_test.go (hunk 2 above)
Comment (kanapuli) [reply opcua_plugin/browse_test.go:179]: No the attribute for `browseName` is created only once through `rootNode.attriburtes = append(..)`.  The "TestNode" in the function parameter is mostly useless and is used only inside the Mock's BrowseName() function(shown in the image). It is just easy to return the name directly instead of looping through the attributes and returning the browsename attribute
image

File: opcua_plugin/browse_test.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+			}
+
+			var errs []error
+			for err := range errChan {
+				errs = append(errs, err)
+			}
+
+			Expect(errs).Should(BeEmpty())
+			Expect(nodes).Should(HaveLen(1))
+		})
```
[the remark this answers is not in the document]
Comment (kanapuli) [reply opcua_plugin/browse_test.go:283]: May be later. Not in this PR

[the remark this answers is not in the document]
Comment (kanapuli) [reply opcua_plugin/browse_test.go]: Done

File: opcua_plugin/browse_test.go (hunk 1 above)
Comment (kanapuli) [reply opcua_plugin/browse_test.go]: Removed

**Review by JeremyTheocharis — APPROVED** [review]
